strategy_infura.go

```diff
@@ -0,0 +1,443 @@
+	newInstances, chanErr := detectNewInstances(ctx, sess, asgName, currentInstances)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-chanErr:
+				fmt.Printf("error while polling ASG details: %v\n", err)
+				continue
+			}
+		}
+	}()
+
+	// make sure to stop the following call when done
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	readyInstances, chanErr2 := detectInstancesReady(subCtx, sess, tg, newInstances)
+
+	instanceReady := make([]instance, 0, count)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-chanErr2:
+			fmt.Printf("error while polling autoscaling or target group reported target health: %v\n", err)
+			continue
+		case inst := <-readyInstances:
+			instanceReady = append(instanceReady, inst)
+			fmt.Printf("\t(%d/%d) Instance %s is ready\n", len(instanceReady), count, inst.instanceId)
+			if len(instanceReady) >= count {
+				return instanceReady, nil
+			}
+		}
+	}
+}
+
+func detectNewInstances(ctx context.Context, sess *session.Session, asgName string, currentInstances []instance) (chan instance, chan error) {
+	oldSet := make(map[string]instance)
+	for _, inst := range currentInstances {
+		oldSet[inst.instanceId] = inst
+	}
+
+	chanOut := make(chan instance)
+	chanErr := make(chan error)
+
+	period := 5 * time.Second
+
+	go func() {
+		defer close(chanOut)
+		defer close(chanErr)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(period):
+			}
+
+			asg, err := getASGDetails(ctx, sess, asgName)
+			if err != nil {
+				chanErr <- err
+				// don't stop on error, let the caller decide what to do
+				continue
+			}
+
+			for _, inst := range asg.instances {
+				if _, has := oldSet[inst.instanceId]; !has {
+					chanOut <- inst
+					oldSet[inst.instanceId] = inst
+				}
+			}
+		}
+	}()
+
+	return chanOut, chanErr
+}
+
+func detectInstancesReady(ctx context.Context, sess *session.Session, tg *targetGroup, newInstances chan instance) (chan instance, chan error) {
+	chanOut := make(chan instance)
+	chanErr := make(chan error)
+
+	period := 10 * time.Second
+
+	go func() {
+		defer close(chanOut)
+		defer close(chanErr)
+
+		instanceSet := make(map[string]instance)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case inst := <-newInstances:
+				fmt.Printf("\tfound new instance: %s\n", inst)
+				instanceSet[inst.instanceId] = inst
+			case <-time.After(period):
+				// wait to have at least one instance to inspect
+				if len(instanceSet) == 0 {
+					continue
+				}
+			}
+
+			// inspect the instance health
+			healthyInstances, err := getHealthyAutoscalingInstances(ctx, sess, instanceSet)
+			if err != nil {
+				chanErr <- err
+				// don't stop on error, let the caller decide what to do
+				continue
+			}
+
+			// if we don't have a target group, end the checks here
```

Comment:
This may be another area to revisit, as we'd still need a target group-style health check to know if the instances are effectively healthy (the ASG health check is separate). For example, if we used this on an emergency hammerhead ASG not tied to any TG, we'd take out the old instances before the new ones are healthy from an eth1 service point of view.

Reply:
so that would be an option to pass an arbitrary health check ? There would be no way to detect that, right?

Reply:
Yeah I think in the future along with custom strategies we may want the ability to compose in a custom health check as well.